File: clients/android/app/src/main/kotlin/com/lextures/android/features/reader/ReaderLogic.kt
package com.lextures.android.features.reader

import com.lextures.android.core.lms.CaptionRecord
import com.lextures.android.core.lms.ReadingPreferencesRow
import java.net.URI
import java.util.Locale

object ReaderLogic {
    data class VttCue(val start: Double, val end: Double, val text: String)

    val commonLocales = listOf(
        "en" to "English",
        "es" to "Spanish",
        "fr" to "French",
        "de" to "German",
        "ar" to "Arabic",
        "zh" to "Chinese",
        "ja" to "Japanese",
        "ko" to "Korean",
    )

    fun parseVtt(raw: String): List<VttCue> {
        val lines = raw.replace("\r\n", "\n").split("\n")
        val cues = mutableListOf<VttCue>()
        var index = 0
        while (index < lines.size) {
            val line = lines[index].trim()
            if (line.isEmpty() || line == "WEBVTT" || line.startsWith("NOTE")) {
                index++
                continue
            }
            if ("-->" in line) {
                val parts = line.split("-->")
                if (parts.size == 2) {
                    val start = parseVttTimestamp(parts[0].trim())
                    val end = parseVttTimestamp(parts[1].trim().substringBefore(" "))
                    if (start != null && end != null) {
                        index++
                        val textLines = mutableListOf<String>()
                        while (index < lines.size && lines[index].trim().isNotEmpty()) {
                            textLines += stripVttTags(lines[index])
                            index++
                        }
                        val text = textLines.joinToString(" ").trim()
                        if (text.isNotEmpty()) cues += VttCue(start, end, text)
                        continue
                    }
                }
            }
            index++
        }
        return cues
    }

    fun activeCue(time: Double, cues: List<VttCue>): VttCue? =
        cues.firstOrNull { time >= it.start && time < it.end }

    fun videoUrl(text: String): String? {
        val trimmed = text.trim()
        if (!trimmed.startsWith("http", ignoreCase = true)) return null
        val uri = runCatching { URI(trimmed) }.getOrNull() ?: return null
        val host = uri.host?.lowercase(Locale.US).orEmpty()
        if (host.contains("youtube.com") || host.contains("youtu.be") || host.contains("vimeo.com")) {
            return trimmed
        }
        val path = uri.path.lowercase(Locale.US)
        if (listOf(".mp4", ".mov", ".m3u8", ".webm").any { path.endsWith(it) }) {
            return trimmed
        }
        return null
    }

    fun storageObjectId(url: String): String? {
        val patterns = listOf(
            Regex("""/api/v1/files/([0-9a-fA-F-]{36})"""),
            Regex("""/files/([0-9a-fA-F-]{36})"""),
        )
        for (pattern in patterns) {
            pattern.find(URI(url).path)?.groupValues?.getOrNull(1)?.let { return it }
        }
        return null
    }

    fun localeLabel(code: String): String =
        commonLocales.firstOrNull { it.first == code.lowercase(Locale.US) }?.second ?: code.uppercase(Locale.US)

    fun readyCaptions(records: List<CaptionRecord>): List<CaptionRecord> =
        records.filter { it.status.equals("ready", ignoreCase = true) }

    fun defaultReadingPreferences(): ReadingPreferencesRow = ReadingPreferencesRow()

    fun dyslexiaFromFontFace(fontFace: String): Boolean = fontFace == "open-dyslexic"

    fun fontFaceFromDyslexia(enabled: Boolean, current: String): String =
        if (enabled) "open-dyslexic" else if (current == "open-dyslexic") "default" else current

    private fun parseVttTimestamp(value: String): Double? {
        val chunks = value.trim().split(":")
        if (chunks.size < 2) return null
        var seconds = 0.0
        if (chunks.size == 3) seconds += chunks[0].toDoubleOrNull()?.times(3600) ?: return null
        val minuteIndex = if (chunks.size == 3) 1 else 0
        val secondIndex = if (chunks.size == 3) 2 else 1
        val minutes = chunks[minuteIndex].toDoubleOrNull() ?: return null
        val secParts = chunks[secondIndex].split(".")
        val secs = secParts[0].toDoubleOrNull() ?: return null
        seconds += minutes * 60 + secs
        if (secParts.size > 1) {
            seconds += (secParts[1].take(3).toDoubleOrNull() ?: 0.0) / 1000.0
        }
        return seconds
    }

    private fun stripVttTags(line: String): String = line.replace(Regex("<[^>]+>"), "")
}

File: clients/ios/Lextures/Features/Reader/ReaderLogic.swift
import Foundation

/// Pure helpers for immersive reader: VTT parsing, prefs, and feature gating (M6.3).
enum ReaderLogic {
    struct VttCue: Equatable {
        var start: TimeInterval
        var end: TimeInterval
        var text: String
    }

    static let commonLocales: [(code: String, label: String)] = [
        ("en", "English"), ("es", "Spanish"), ("fr", "French"), ("de", "German"),
        ("ar", "Arabic"), ("zh", "Chinese"), ("ja", "Japanese"), ("ko", "Korean"),
    ]

    static func parseVtt(_ raw: String) -> [VttCue] {
        let lines = raw
            .replacingOccurrences(of: "\r\n", with: "\n")
            .split(separator: "\n", omittingEmptySubsequences: false)
            .map(String.init)

        var cues: [VttCue] = []
        var index = 0
        while index < lines.count {
            let line = lines[index].trimmingCharacters(in: .whitespaces)
            if line.isEmpty || line == "WEBVTT" || line.hasPrefix("NOTE") {
                index += 1
                continue
            }
            if line.contains("-->") {
                let parts = line.components(separatedBy: "-->")
                guard parts.count == 2,
                      let start = parseVttTimestamp(parts[0].trimmingCharacters(in: .whitespaces)),
                      let end = parseVttTimestamp(parts[1].trimmingCharacters(in: .whitespaces).components(separatedBy: " ").first ?? "")
                else {
                    index += 1
                    continue
                }
                index += 1
                var textLines: [String] = []
                while index < lines.count, !lines[index].trimmingCharacters(in: .whitespaces).isEmpty {
                    textLines.append(stripVttTags(lines[index]))
                    index += 1
                }
                let text = textLines.joined(separator: " ").trimmingCharacters(in: .whitespacesAndNewlines)
                if !text.isEmpty {
                    cues.append(VttCue(start: start, end: end, text: text))
                }
                continue
            }
            index += 1
        }
        return cues
    }

    static func activeCue(at time: TimeInterval, in cues: [VttCue]) -> VttCue? {
        cues.first { time >= $0.start && time < $0.end }
    }

    static func storageObjectId(from url: URL) -> String? {
        let path = url.path
        let patterns = [
            #"/api/v1/files/([0-9a-fA-F-]{36})"#,
            #"/files/([0-9a-fA-F-]{36})"#,
        ]
        for pattern in patterns {
            if let match = path.range(of: pattern, options: .regularExpression) {
                let segment = String(path[match])
                if let id = segment.split(separator: "/").last {
                    return String(id)
                }
            }
        }
        return nil
    }

    static func localeLabel(_ code: String) -> String {
        commonLocales.first { $0.code == code.lowercased() }?.label ?? code.uppercased()
    }

    static func readyCaptions(_ records: [CaptionRecord]) -> [CaptionRecord] {
        records.filter { $0.status.lowercased() == "ready" }
    }

    static func defaultReadingPreferences() -> ReadingPreferencesRow {
        ReadingPreferencesRow()
    }

    static func mergeReadingPreferences(
        local: ReadingPreferencesRow,
        server: ReadingPreferencesRow
    ) -> ReadingPreferencesRow {
        var merged = server
        if local.updatedAt != nil, server.updatedAt == nil {
            merged = local
        }
        return merged
    }

    static func dyslexiaFromFontFace(_ fontFace: String) -> Bool {
        fontFace == "open-dyslexic"
    }

    static func fontFaceFromDyslexia(_ enabled: Bool, current: String) -> String {
        enabled ? "open-dyslexic" : (current == "open-dyslexic" ? "default" : current)
    }

    // MARK: - Private

    private static func parseVttTimestamp(_ value: String) -> TimeInterval? {
        let cleaned = value.trimmingCharacters(in: .whitespaces)
        let chunks = cleaned.split(separator: ":").map(String.init)
        guard chunks.count >= 2 else { return nil }
        var seconds: Double = 0
        if chunks.count == 3, let hours = Double(chunks[0]) {
            seconds += hours * 3600
        }
        let minuteIndex = chunks.count == 3 ? 1 : 0
        let secondIndex = chunks.count == 3 ? 2 : 1
        guard let minutes = Double(chunks[minuteIndex]),
              let secParts = chunks[secondIndex].split(separator: ".").map(String.init) as [String]?,
              let secs = Double(secParts[0])
        else { return nil }
        seconds += minutes * 60 + secs
        if secParts.count > 1, let millis = Double(String(secParts[1].prefix(3))) {
            seconds += millis / 1000
        }
        return seconds
    }

    private static func stripVttTags(_ line: String) -> String {
        line.replacingOccurrences(of: #"<[^>]+>"#, with: "", options: .regularExpression)
    }
}